examples/layer_shell.rs

```diff
@@ -0,0 +1,255 @@
+use smithay_client_toolkit::{
+    environment,
+    environment::{Environment, SimpleGlobal},
+    output::{with_output_info, OutputHandler, OutputHandling, OutputInfo, OutputStatusListener},
+    reexports::{
+        calloop,
+        client::protocol::{wl_compositor, wl_output, wl_shm, wl_surface},
+        client::{Attached, Display, Main, Proxy},
+        protocols::wlr::unstable::layer_shell::v1::client::{
+            zwlr_layer_shell_v1, zwlr_layer_surface_v1,
+        },
+    },
+    shm::{DoubleMemPool, ShmHandler},
+    WaylandSource,
+};
+
+use byteorder::{NativeEndian, WriteBytesExt};
+
+use std::cell::{Cell, RefCell};
+use std::io::{BufWriter, Seek, SeekFrom, Write};
+use std::rc::Rc;
+
+struct Env {
+    compositor: SimpleGlobal<wl_compositor::WlCompositor>,
+    layer_shell: SimpleGlobal<zwlr_layer_shell_v1::ZwlrLayerShellV1>,
+    shm: ShmHandler,
+    outputs: OutputHandler,
+}
+
+impl OutputHandling for Env {
+    fn listen<F: FnMut(wl_output::WlOutput, &OutputInfo) + 'static>(
+        &mut self,
+        f: F,
+    ) -> OutputStatusListener {
+        self.outputs.listen(f)
+    }
+}
+
+environment!(Env,
+    singles = [
+        wl_compositor::WlCompositor => compositor,
+        zwlr_layer_shell_v1::ZwlrLayerShellV1 => layer_shell,
+        wl_shm::WlShm => shm,
+    ],
+    multis = [
+        wl_output::WlOutput => outputs
+    ]
+);
```

Comment:
Unless I'm missing some subtlety, declaring the `Env` struct could be done using the `default_environment!()` macro with layer shell as an addon, like so:

```rust
default_environment!(Env,
    fields = [
        layer_shell: SimpleGlobal<zwlr_layer_shell_v1::ZwlrLayerShellV1>,
    ],
    singles = [
                 zwlr_layer_shell_v1::ZwlrLayerShellV1 => layer_shell
    ],
)
```

rather than manually creating the struct and its implementations.

Reply:
Yeah that should be totally fine.